services/src/unit-test/java/uk/gov/moj/sdt/services/utils/SdtBulkReferenceGeneratorTest.java

```diff
@@ -0,0 +1,76 @@
+package uk.gov.moj.sdt.services.utils;
+
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoExtension;
+
+import uk.gov.moj.sdt.dao.GenericDao;
+import uk.gov.moj.sdt.dao.api.IGenericDao;
+import uk.gov.moj.sdt.services.utils.api.ISdtBulkReferenceGenerator;
+import uk.gov.moj.sdt.utils.AbstractSdtUnitTestBase;
+
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.mockito.Mockito.mock;
+
+
+@ExtendWith(MockitoExtension.class)
+public class SdtBulkReferenceGeneratorTest extends AbstractSdtUnitTestBase{
+    @Mock
+    IGenericDao genericDaoMock;
+
+    private static final String FOUR_CHARACTERS_ONLY ="The target application length is expected to be 4 characters.";
+
+    ISdtBulkReferenceGenerator sdtBulkReferenceGenerator;
+    @BeforeEach
+    @Override
+    public void setUp() {
+
+        sdtBulkReferenceGenerator = new SdtBulkReferenceGenerator(genericDaoMock);
+
+    }
+
+    @Test
+    void testGetSdtBulkReferenceGenerator() {
+
+        String generatedRef = sdtBulkReferenceGenerator.getSdtBulkReference("MCOL");
+
+        assertNotNull(generatedRef);
+    }
```

Comment:
Test needs to set up the expected return value for genericDaoMock.getNextSequenceValue().  This can then be used when checking the returned generated reference.  Don't forget to verify if afterwards.

The reference contains the current time, so to ensure it has a consistent value set up a static mock for ```System.currentTimeMillis()```.  The test can then construct the expected reference from the system id (MCOL), the expected time (in the format yyyyMMddHHmmss) and the sequence value set up to be returned by getNextSequenceValue() and use assertEquals() to compare it with what is returned.

Reply:
Getting an error stating that mocking java.lang.System methods will cause issues with class loading.
Have mocked the getNextSequenceValue and verified it

Reply:
Apologies - wasn't aware of that restriction.

The new version of the test is a little mixed up - the mock behaviour for the genericDao object is initially being set after the getSdtBulkReference() method has been called, then the getSdtBulkReference() method is being called again in assertTrue().  The test could also use a regular expression to check the returned reference.  Try changing it to:

```
when(genericDaoMock.getNextSequenceValue("SDT_REF_SEQ")).thenReturn(999L);
String generatedRef = sdtBulkReferenceGenerator.getSdtBulkReference("MCOL");

assertNotNull(generatedRef);
assertTrue(generatedRef.matches("^MCOL-[\\d]{14}-000000999$"), "Reference does not match expected pattern");

verify(genericDaoMock).getNextSequenceValue("SDT_REF_SEQ");
```
We know the first and last parts of the reference as those are being passed in and set as the value returned by the mock genericDao.  As the middle part will change depending on the current date and time we can just check that it's 14 digits.

Reply:
done